jormungandr/src/fragment/logs.rs

```diff
@@ -1,211 +1,137 @@
-                .map(|log| self.insert(log))
-                .filter(|was_modified| *was_modified)
-                .count()
-        }
-
-        pub fn modify(&mut self, fragment_id: &Hash, status: FragmentStatus) {
-            let len = self.entries.len();
-            match self.entries.entry(fragment_id.clone()) {
-                Entry::Occupied(mut entry) => {
-                    entry.get_mut().0.modify(status);
-
-                    self.expirations
-                        .reset_at(&entry.get().1, Instant::now() + self.ttl);
-                }
-                Entry::Vacant(entry) => {
-                    // while a log modification, if the log was not already present in the
-                    // logs it means we received it from the a new block from the network.
-                    // we can mark the status of the transaction so newly received transaction
-                    // be stored.
-
-                    if self.max_entries < len {
-                        let delay = self.expirations.insert(*fragment_id, self.ttl);
-                        entry.insert((
-                            FragmentLog::new(
-                                fragment_id.clone().into_hash(),
-                                FragmentOrigin::Network,
-                            ),
-                            delay,
-                        ));
-                    }
+    fn poll_purge_internal(&mut self, cx: &mut Context) -> Poll<Result<(), time::Error>> {
+        loop {
+            match self.expirations.as_mut().poll_next(cx) {
+                Poll::Ready(Some(Ok(entry))) => {
+                    self.entries.remove(entry.get_ref());
                 }
-            }
-        }
-
-        pub fn poll_purge(&mut self, cx: &mut Context) -> Poll<Result<(), time::Error>> {
-            loop {
-                match self.expirations.as_mut().poll_next(cx) {
-                    Poll::Ready(Some(Ok(entry))) => {
-                        self.entries.remove(entry.get_ref());
-                    }
-                    Poll::Ready(Some(Err(e))) => return Poll::Ready(Err(e)),
-                    Poll::Ready(None) => return Poll::Ready(Ok(())),
+                Poll::Ready(Some(Err(e))) => return Poll::Ready(Err(e)),
+                Poll::Ready(None) => return Poll::Ready(Ok(())),
 
-                    // Here Pending means there are still items in the DelayQueue but
-                    // they are not expired. We don't want this function to wait for these
-                    // ones to expired. We only cared about removing the expired ones.
-                    Poll::Pending => return Poll::Ready(Ok(())),
-                }
+                // Here Pending means there are still items in the DelayQueue but
+                // they are not expired. We don't want this function to wait for these
+                // ones to expired. We only cared about removing the expired ones.
+                Poll::Pending => return Poll::Ready(Ok(())),
```

Comment:
> so that the GC task is just an async loop around calling `purge` without a periodic timer

If you mean a stream of expired GC entries combined with a stream of requests coming to the process, then it makes sense to me. Will let us to get rid of a magic number of seconds at least. If you mean spawning such loop in as a separate process, it will require to lock `entries`, which is questionable.

> We might need a non-futuristic, purely synchronous equivalent of DelayQueue for all these GC things.

I think we need to think of how does it compare to the async version in terms of performance. Without digging into that both sound like good ideas that are just different.